app/models/pusher.rb

```diff
@@ -196,18 +194,20 @@ def republish_notification(version)
              "Please bump the version number and push a new different release.\n" \
              "See also `gem yank` if you want to unpublish the bad release.", 409)
     else
-      different_owner = "pushed by a previous owner of this gem " unless version.rubygem.owners.include?(@user)
+      different_owner = "pushed by a previous owner of this gem " unless owner.owns_gem?(version.rubygem)
       notify("A yanked version #{different_owner}already exists (#{version.full_name}).\n" \
              "Repushing of gem versions is not allowed. Please use a new version and retry", 409)
     end
   end
 
   def notify_unauthorized
-    if rubygem.unconfirmed_ownership?(user)
+    if !api_key.user?
+      notify("You are not allowed to push this gem.", 403)
+    elsif rubygem.unconfirmed_ownership?(owner)
       notify("You do not have permission to push to this gem. " \
-             "Please confirm the ownership by clicking on the confirmation link sent your email #{user.email}", 403)
+             "Please confirm the ownership by clicking on the confirmation link sent your email #{owner.email}", 403)
     else
-      notify("You do not have permission to push to this gem. Ask an owner to add you with: gem owner #{rubygem.name} --add #{user.email}", 403)
+      notify("You do not have permission to push to this gem. Ask an owner to add you with: gem owner #{rubygem.name} --add #{owner.email}", 403)
```

Comment:
Unrelated, but does this mean we leak emails here?

Reply:
no, it's the users own email

Reply:
Oh, that makes sense.